### DISCUSSION TITLE
hls4ml installation

### DISCUSSION THREAD

**Discussion by amassahb** [fastmachinelearning/hls4ml#966] [General]
Hi to all,
I am new here. I just started to install hls4ml library by running the following command in my home folder:
**pip3 install hls4ml**
but I am getting the following error:

<pre>Collecting hls4ml
  Using cached https://files.pythonhosted.org/packages/e9/64/30fe14f8ce47277c2a69d48a63538ed5bd54241dec388048d457e5cf4df3/hls4ml-0.6.0-py3-none-any.whl
Requirement already satisfied: pyyaml in /usr/lib64/python3.6/site-packages (from hls4ml)
Requirement already satisfied: numpy in /usr/lib64/python3.6/site-packages (from hls4ml)
Collecting onnx&gt;=1.4.0 (from hls4ml)
  Using cached https://files.pythonhosted.org/packages/8f/71/1543d8dad6a26df1da8953653ebdbedacea9f1a5bcd023fe10f8c5f66d63/onnx-1.14.1.tar.gz
    Complete output from command python setup.py egg_info:
    fatal: Not a git repository (or any parent up to mount point /tmp)
    Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
    Traceback (most recent call last):
      File &quot;&lt;string&gt;&quot;, line 1, in &lt;module&gt;
      File &quot;/tmp/pip-build-svy0ktzx/onnx/setup.py&quot;, line 310, in &lt;module&gt;
        packages = setuptools.find_packages() + setuptools.find_namespace_packages(
    AttributeError: module &apos;setuptools&apos; has no attribute &apos;find_namespace_packages&apos;
    
    ----------------------------------------
<font color="#CC0000">Command &quot;python setup.py egg_info&quot; failed with error code 1 in /tmp/pip-build-svy0ktzx/onnx/</font></pre>

Do you have any idea?

Thank you in advance for your support.

Regards,
Amir

**Comment by vloncar** [comment]
You're using an old version of python (3.6), so you're attempting to install a last version of hls4ml that supported that python. You should upgrade to at least python 3.8 but preferably newer. We'll be moving to 3.10 as the requirement soon.

**Comment by amassahb** [comment]
Thank you for your reply. I am quite new with Python world.
I finally could install it without any error. However, when I try to import it I get the following errors:

`import hls4ml
WARNING:tensorflow:From C:\Program Files\Python310\lib\site-packages\keras\src\losses.py:2976: The name tf.losses.sparse_softmax_cross_entropy is deprecated. Please use tf.compat.v1.losses.sparse_softmax_cross_entropy instead.

Traceback (most recent call last):
  File "<pyshell#0>", line 1, in <module>
    import hls4ml
  File "C:\Program Files\Python310\lib\site-packages\hls4ml\__init__.py", line 1, in <module>
    from hls4ml import converters, report, utils  # noqa: F401
  File "C:\Program Files\Python310\lib\site-packages\hls4ml\converters\__init__.py", line 7, in <module>
    from hls4ml.converters.keras_to_hls import KerasFileReader  # noqa: F401
  File "C:\Program Files\Python310\lib\site-packages\hls4ml\converters\keras_to_hls.py", line 5, in <module>
    from hls4ml.model import ModelGraph
  File "C:\Program Files\Python310\lib\site-packages\hls4ml\model\__init__.py", line 1, in <module>
    from hls4ml.model.graph import HLSConfig, ModelGraph  # noqa: F401
  File "C:\Program Files\Python310\lib\site-packages\hls4ml\model\graph.py", line 9, in <module>
    from hls4ml.backends import get_backend
  File "C:\Program Files\Python310\lib\site-packages\hls4ml\backends\__init__.py", line 1, in <module>
    from hls4ml.backends.backend import Backend, get_available_backends, get_backend, register_backend  # noqa: F401
  File "C:\Program Files\Python310\lib\site-packages\hls4ml\backends\backend.py", line 5, in <module>
    from hls4ml.backends.template import Template
  File "C:\Program Files\Python310\lib\site-packages\hls4ml\backends\template.py", line 1, in <module>
    from hls4ml.model.optimizer.optimizer import OptimizerPass
  File "C:\Program Files\Python310\lib\site-packages\hls4ml\model\optimizer\__init__.py", line 4, in <module>
    from hls4ml.model.optimizer.optimizer import (  # noqa: F401
  File "C:\Program Files\Python310\lib\site-packages\hls4ml\model\optimizer\optimizer.py", line 5, in <module>
    from hls4ml.utils.string_utils import convert_to_snake_case
  File "C:\Program Files\Python310\lib\site-packages\hls4ml\utils\__init__.py", line 1, in <module>
    from hls4ml.utils.config import config_from_keras_model, config_from_onnx_model, config_from_pytorch_model  # noqa: F401
  File "C:\Program Files\Python310\lib\site-packages\hls4ml\utils\config.py", line 3, in <module>
    import qkeras
  File "C:\Program Files\Python310\lib\site-packages\qkeras\__init__.py", line 23, in <module>
    from .estimate import *  # pylint: disable=wildcard-import
  File "C:\Program Files\Python310\lib\site-packages\qkeras\estimate.py", line 41, in <module>
    from .qlayers import QActivation
  File "C:\Program Files\Python310\lib\site-packages\qkeras\qlayers.py", line 55, in <module>
    from .quantizers import *
  File "C:\Program Files\Python310\lib\site-packages\qkeras\quantizers.py", line 29, in <module>
    from .safe_eval import safe_eval
  File "C:\Program Files\Python310\lib\site-packages\qkeras\safe_eval.py", line 21, in <module>
    from pyparsing import delimitedList
ModuleNotFoundError: No module named 'pyparsing'`

**Comment by amassahb** [comment]
Hi, 
I found the reason and solved it :). Maybe it is worth it to give some information abut my platform. I installed hls4ml on Windows 11, Python 3.10.10, Vitis_HLS 2023.2.
I finally could install it and tried to run this example code which is given in your website:

```
import hls4ml
config = hls4ml.utils.fetch_example_model('KERAS_3layer.json')
print(config)
hls_model = hls4ml.converters.keras_to_hls(config)
hls_model.compile()
```
I can see that the model is created successfully, but I get an error for compiling the model which is:

`'bash' is not recognized as an internal or external command,
operable program or batch file.
Traceback (most recent call last):
  File "c:\hls4ml\test.py", line 11, in <module>
    hls_model.compile()
  File "C:\Program Files\Python310\lib\site-packages\hls4ml\model\graph.py", line 662, in compile
    self._compile()
  File "C:\Program Files\Python310\lib\site-packages\hls4ml\model\graph.py", line 665, in _compile
    lib_name = self.config.backend.compile(self)
  File "C:\Program Files\Python310\lib\site-packages\hls4ml\backends\fpga\fpga_backend.py", line 131, in compile
    raise Exception(f'Failed to compile project "{model.config.get_project_name()}"')
Exception: Failed to compile project "myproject"`

It seems the problem is in sth related to bash files or sth like that. 

I would be really grateful if you could help me or give me some hints dude.

Thank you so much and looking forward to hearing from you.

Regards,
Amir

**Comment by vloncar** [comment]
Here's a hint dude, hls4ml doesn't work on Windows (see the docs). You'll have to use Linux (via VM or WSL). Sorry :-(

**Reply by amassahb** [reply]
Ok, I installed an Ubuntu for that, Ubuntu 22.4 with python 3.10.12. I cold install the hls4ml without any error. I can see it by running:

`pip show hls4ml`

But when I import it into my code and run it, the import line code leads to a crash and error is "Illegal instruction (core dumped)".
This is my code with only a print and the import line:

`print("#######################################################Test1")
import hls4ml
print("Test2")`

I tried to run it in a verbose mode (by using this command "python -v test.py") to see if there is any error, but I cannot see any. I have attached the log file.
[log.txt](https://github.com/fastmachinelearning/hls4ml/files/14466119/log.txt)

At the beginning I though it might be because of Python version, and I tried with python 3.9.18 but not difference.

I would appreciate any support.

Sincerely yours,
Amir

**Reply by amassahb** [reply]
One more point. I have installed my Ubuntu on the VirtualBox. Could it be the reason?

**Reply by jmitrevs** [reply]
It seems like there's an issue with your tensorflow installation. I tend to setup my python environment with miniforge with the following packages:

conda create -n fastml python=3.10
conda activate fastml
conda install tensorflow pytorch numpy matplotlib scikit-learn pandas pytest pytest-cov jupyterlab toposort clize seaborn pydot

And then pip install hls4ml (not conda install).

**Reply by brtsfr** [reply]
hello amassahb, how was your problem? did you solve it then?

**Comment by brtsfr** [comment]
hi i'm new learner here. i also encounted the same situation. i am running ubuntu24.04 in virtualbox of Win11. "pip show hls4ml" indicated a successful installation. but when i typed “hls4ml -h” in cmd, or "import hls4ml" in python, only got "illegal instruction". please help

**Reply by amassahb** [reply]
Hi @brtsfr ,

Yes, thanks to @jmitrevs , I was able to solve my problem. Please follow his post on March 2.

Good luck

**Reply by brtsfr** [reply]
yes, thanks! after several attempts i managed to solve it. i encounted "no module named tf_keras", and simply use "pip install tf_keras".

**Reply by Rezef** [reply]
I am trying to install Xilinx Vivado 2019.2 , but I see that the Vivado archive has an All OS support  archive file which when extracted contains a Windows executible file, how do you get to install the file in your VM?

**Reply by brtsfr** [reply]
put it into your VM first, and then extract it.

**Reply by Rezef** [reply]
I downloaded the file in  my VM but instead of getting a .bin setup file , I got a  windows executible file.

**Reply by Rezef** [reply]
Can you tell me which version of VIvado did you download?

**Reply by brtsfr** [reply]
[Vivado HLx 2019.2: All OS installer Single-File Download ](https:/[/]www.xilinx.com/member/forms/download/xef.html?filename=Xilinx_Vivado_2019.2_1106_2127.tar.gz)(TAR/GZIP - 26.55 GB)

MD5 SUM Value : e2b2762964ef5f014591b13d77d823ab

on page: https:/[/]www.xilinx.com/support/download/index.html/content/xilinx/en/downloadNav/vivado-design-tools/archive.html

**Reply by Rezef** [reply]
Thanks a lot, managed to run the first tutorial notebook completely

**Reply by bluemoon0321** [reply]
hello Rezef, may I ask which tutorial did you finish?
And did you use it on a ubuntu 24.04 and did these steps after cloning to setup the env?
![image](https://github.com/user-attachments/assets/191e83fa-d8db-40b2-88a5-f93f2b5f6c02)

Thanks!

**Reply by Rezef** [reply]
@bluemoon0321 I was talking about this tutorial repo:
https://github.com/fastmachinelearning/hls4ml-tutorial
I installed this on ubuntu 22.02. You just need to create an environment based on the environment.yml mentioned in the tutorial repo. But you should have a Vivado(2020.1) or Vitis backend(23.2) installed in your local machine or virtual machine for this to work

**Reply by bluemoon0321** [reply]
@Rezef oh ok got it. 
One more question .Did you use Vivado through the whole tutorial? because I have vitis(24.2) installed and can successful run the first few cells in part 1, but I see that for part 7 you need to use Vivado. I'm not sure do I also need to install Vivado(2019.2/2020.1) or can I do it through vitis installed only.
Thanks a lot!

**Reply by bo3z** [reply]
For part 7 you currently need access to Vivado HLS. In the future, there will be support to run part 7 using Vitis HLS, as there is a draft PR for that currently.
